nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+// VALUE: Parsed amount to attach
+
+interface AssetIntentOverrides {
+  inputs: AssetInput[];
+  outputs: AssetOutput[];
+}
+
+interface AssetInput {
+  txid: string;
+  index: number;
+}
+
+interface AssetOutput {
+  asset: string;
+  address: number;
+  value: string;
+}
+
+</pre>
+
+
+'''Success return value'''
+
+interface InvokeOutput {
+  txid: string; // The transaction ID of the invocation
+  nodeUrl?: string; // The node which the transaction was broadcast to. Returned if transaction is broadcast by wallet provider
+  signedTx?: string; // The serialized signed transaction. Only returned if the broadcastOverride input argument was set to True
+}
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`SEND_ERROR`|`MALFORMED_INPUT`|`CANCELED`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+invoke({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  operation: 'storeData',
+  arguments: [
+    {
+      type: 'string',
+      value: 'hello'
+    }
+  ],
+  attachedAssets: {
+    NEO: '100',
+    GAS: '0.0001',
+  },
+  fee: '0.001'
+})
+.then(({txid, nodeUrl}: InvokeOutput) => {
+  console.log('Invoke transaction success!');
+  console.log('Transaction ID: ' + txid);
+  console.log('RPC node URL: ' + nodeUrl);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case RPC_ERROR:
+      console.log('There was an error when broadcasting this transaction to the network.');
+      break;
+    case CANCELED:
+      console.log('The user has canceled this transaction.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hash is valid
+** Format parameters
+*** If `byteArray`, convert to appropriate hexstring based on assumed format
+* Set script transaction attribute 0x20 according to the following conditions
+** If triggerContractVerification is set to true, set 0x20 to scriptHash of the contract being invoked
+** If there is no fee, attachedAssets, or 'assetIntentOverrides', set 0x20 to the users address
+** If there are assetIntentOverrides but none of the inputs belong to the user address, set 0x20 to user address
+* If the "broadcastOverride" input argument is set to True, return the serialized signed transaction to the dApp
+* Else broadcast the RPC request
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}:
+</pre>
+
+
+====invokeMulti====
+Execute a multiple contract invocations in a single transaction.
+
+
+'''Method Interface'''
+
+<pre>
+function invokeMulti({
+  invokeArgs: InvokeArguments[],
+  fee?: string,
```

Comment:
which fee? sys? net?

Reply:
I just saw some text on `deploy` section, saying that "costs are calculated by provider and shown to user". So perhaps this is just network fee? Anyway, even network fee can be calculated by provider too... so this `fee` may perhaps become a struct `fee = { netfee: ... sysfee: ...}` returned by provider, and accepted by user.

Reply:
yes, this is the network fee.